drivers/dai/intel/dmic/dmic_nhlt.c

```diff
@@ -125,6 +126,68 @@ static int dai_nhlt_dmic_dai_params_get(struct dai_intel_dmic *dmic,
 
 	return 0;
 }
+
+
+/*
+ * @brief Set clock source used by device
+ *
+ * @param source Clock source index
+ */
+static inline void dai_dmic_clock_select_set(const struct dai_intel_dmic *dmic, uint32_t source)
+{
+	uint32_t val;
+#ifdef CONFIG_SOC_INTEL_ACE20_LNL /* Ace 2.0 */
+	val = sys_read32(dmic->vshim_base + DMICLVSCTL_OFFSET);
+	val &= ~DMICLVSCTL_MLCS;
+	val |= FIELD_PREP(DMICLVSCTL_MLCS, source);
+	sys_write32(val, dmic->vshim_base + DMICLVSCTL_OFFSET);
+#else
+	val = sys_read32(dmic->shim_base + DMICLCTL_OFFSET);
+	val &= ~DMICLCTL_MLCS;
+	val |= FIELD_PREP(DMICLCTL_MLCS, source);
+	sys_write32(val, dmic->shim_base + DMICLCTL_OFFSET);
+#endif
+}
+
+/*
+ * @brief Get clock source used by device
+ *
+ * @return Clock source index
+ */
+static inline uint32_t dai_dmic_clock_select_get(const struct dai_intel_dmic *dmic)
+{
+	uint32_t val;
+#ifdef CONFIG_SOC_INTEL_ACE20_LNL /* Ace 2.0 */
+	val = sys_read32(dmic->vshim_base + DMICLVSCTL_OFFSET);
+	return FIELD_GET(DMICLVSCTL_MLCS, val);
+#else
+	val = sys_read32(dmic->shim_base + DMICLCTL_OFFSET);
+	return FIELD_GET(DMICLCTL_MLCS, val);
+#endif
+}
+
+/*
+ * @brief Set clock source used by device
+ *
+ * @param source Clock source index
+ */
+static int dai_dmic_set_clock(const struct dai_intel_dmic *dmic, const uint8_t clock_source)
+{
+	LOG_DBG("%s(): clock_source = %u", __func__, clock_source);
+
+	if (!adsp_clock_source_is_supported(clock_source)) {
+		return -ENOTSUP;
+	}
+
+#ifndef CONFIG_SOC_INTEL_ACE20_LNL /* Ace 2.0 */
+	if (clock_source && !(sys_read32(dmic->shim_base + DMICLCAP_OFFSET) & DMICLCAP_MLCS)) {
```

Comment:
that test makes no sense to me. Clock source can be zero.

#define ADSP_CLOCK_SOURCE_XTAL_OSC		0

That would prevent us from setting the clock as xtal. Maybe it works because the xtal is the default but that's working by accident and not by design.

Reply:
This condition is not responsible for setting the clock source. It checks the capability register to see if the hardware supports other clock sources.